### PR TITLE
docs: Add source rewriter design doc

### PR DESCRIPTION
**Pull request by ayrtonm**
For reference, the latest function pointer macros are defined in ia2.h [in this PR](https://github.com/immunant/IA2-Phase2/pull/121/files#diff-f732e46bfbfd1583f6055e79a50dc0c290bd90c0a89ada1a2b4101f5bcc198c5).

This makes #158, #131, #130, ~~#51,~~ #49, #43, ~~#24, #22,~~ #21 and #4 obsolete.

edit: #51, #24 and #22 are still relevant

### CODE REVIEW

File: docs/source_rewriter.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+super low priority.
+
+### Is a compile commands JSON required?
+
+Up for debate so it's TBD
+
+### What if I want to make the source file changes myself? 
+
+We can add a flag to only generate the output `.c` with the call gate wrappers
+and leave the input `.c`s unmodified.
```
[the remark this answers is not in the document]
Comment (ayrtonm) [reply docs/source_rewriter.md:195]: Since we're going with replacing fn ptrs again, this is an option but it's low priority.

**Comment by ayrtonm** [comment]
I omitted talking about rewriting function pointer types in the last commit to focus on the new things the tool will rewrite and make it easy to review. We will rewrite function pointer types though.

File: docs/source_rewriter.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+The rewriter will take all source files as input along with which compartment
+each belongs to. It will generate one output `.c` and one `.h` for use with all
+compartments. It will also generate compartment-specific linker args and
+`objcopy --redefine-syms` files.
+
+### Define call gates for direct transitions
+
+The rewriter will work in two phases. In the first phase it'll parse all source
+files and keep track of functions defined and their compartment, functions
+called and from which compartment (TODO: Should we check functions called or
```
Comment (ayrtonm) [inline docs/source_rewriter.md:58]: If we create call gates for undefined functions in each translation unit, then we will almost certainly have more call gates than we actually need but we'll only need to rerun the rewriter when more functions are declared (e.g. a new #include is added).

If we create call gates for functions that are called, then we'll have the minimum number of call gates we need, but we'll need to rerun the rewriter each time a new function call to another compartment is added.

Comment (ayrtonm) [reply docs/source_rewriter.md:58]: Went with the first option in #182 for now.

File: docs/source_rewriter.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+compartment (TODO: What if a user forgets an input .c? maybe assert that these
+functions are declared in /usr/ headers).
+
+In phase two we'll generate the call gates for each direct transition. Since we
+know all the caller and target compartment we can hard-code the pkey in each
+call gate wrapper. These wrappers are inserted with `ld --wrap` so their names
+must start with `__wrap_`. Since a function may be called from multiple
+compartments and call gates have a hardcoded caller pkey, we need one wrapper
+per caller compartment. To distinguish wrappers for different compartments we'll
+name the wrappers `__wrap_$FUNCTION_NAME_from_$CALLER_PKEY`. This will require
```
Comment (rinon) [inline docs/source_rewriter.md:70]: What happened to defining these symbols with hidden visibility? I.e. define `__wrap_foo` separately in compartment1 and compartment2, each with hidden visibility then linking that into the DSO?

File: docs/source_rewriter.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+This wrapper will be identical to the wrappers for external linkage except that
+it will not have `.global __ia2_fn_name` to make the wrapper static. This will
+make it easier to support the case where static functions in different files
+with the same name have their address taken. TODO: We write the
+`IA2_DEFINE_STATIC_WRAPPER_*` functions to one output header so we'd need a way
+to distinguish different files.
+
+### Call gates for indirect callsites
+
+We also need call gates for indirect callsites. Each compartment might not
```
Comment (rinon) [inline docs/source_rewriter.md:151]: Do we rewrite every indirect callsite? Is there a way to opt out for hot callsites known to only call functions internal to its own compartment?

**Pull request merged**